### PR TITLE
Gvaihir/vulcan dex

### PR DESCRIPTION
**Pull request by Gvaihir**
@dtm2451 - please pay attention to workflow scripts and a wrapper class DataIO in archimedes. If you like how it works, feel free to refactor your scripts with regards to read/write of h5. Also `dge.cwl`.
@coleshaw - `SingleDropdownMulticheckbox` is WIP but I could not deploy anything to do the tests and figure out where data IO is happening.

### CODE REVIEW

**Comment by coleshaw** [comment]
I think I have the UI widget working, will write some specs and put out a PR on top of this branch for review.

File: archimedes/archimedes/functions/dataflow/path.py [hunk 1]
```diff
@@ -45,5 +45,7 @@ def output_tsv(data, name, outputs_env=_os.environ, outputs_dir=None):
     return data.to_csv( output_path(name, outputs_env, outputs_dir) , sep='\t')
 
 def output_json(data, name, outputs_env=_os.environ, outputs_dir=None):
+    if not isinstance(data, dict):
```
[the remark this answers is not in the document]
Comment (corps) [reply archimedes/archimedes/functions/dataflow/path.py:48]: I do agree that we need more full end to end tests running.  I think it's possible to do, but I am not clear yet what we should run precisely to make these stronger guarantees, especially also while the codebase is under so much constant change and tests have adjustments frequently anyways.

We could also start adding more typing into the python codebase that helps catch when types are forced or broken.

[the remark this answers is not in the document]
Comment (Gvaihir) [reply archimedes/archimedes/functions/dataflow/path.py:48]: ye, strong assumption, will scratch that

[the remark this answers is not in the document]
Comment (dtm2451) [reply archimedes/archimedes/functions/dataflow/path.py]: We could def work in some more typing expectations.  Best ways to do so in python... `input: str`-like typing in functions definitions when a single type would be assured, and maybe add a `type` input to `input/output_json()` and similar bits for adding (optional) flexible type-assertion?

[the remark this answers is not in the document]
Comment (Gvaihir) [reply archimedes/archimedes/functions/dataflow/path.py]: Python is not a strongly typed language, so those types are not forced, interpreter will still run your code - only IDE will complain when the type is mismatched (let alone the mismatch has to be detected). My packages are typed, but that's just my pref from Scala experience + an attempt to make code maintainability a bit easier.

I reverted that type check

[the remark this answers is not in the document]
Comment (dtm2451) [reply archimedes/archimedes/functions/dataflow/path.py]: Ahh.  well, we could use assertions to enforce typing manually.  That's what I'd been thinking for the `type` input idea implementation.

**Pull request closed without merging**